scripts/target.js

```diff
@@ -0,0 +1,281 @@
+ * @param sessionId
+ * @param useProxy Whether to use the proxy.
+ * @returns {Promise<any>}
+ */
+async function fetchOffers(client, sessionId, useProxy) {
+  const url = `${window.location.protocol}//${window.location.host}`;
+
+  console.debug(`Loading offers for client ${client} and url ${url}`); // eslint-disable-line no-console
+
+  const payload = {
+    context: {
+      channel: 'web',
+      address: {
+        url,
+      },
+    },
+    execute: {
+      pageLoad: {},
+    },
+  };
+
+  const options = {
+    method: 'POST',
+    headers: {
+      'Content-Type': 'application/json',
+      'cache-control': 'no-cache',
+    },
+    body: JSON.stringify(payload),
+  };
+
+  const host = useProxy ? '/' : `https://${client}.tt.omtrdc.net/`;
+  console.debug(`Using target host: ${host}`); // eslint-disable-line no-console
+  const response = await fetch(`${host}rest/v1/delivery?client=${client}&sessionId=${sessionId}`, options);
+  if (!response.ok) {
+    throw new Error(`Failed to fetch offers: ${response.status} ${response.statusText}`);
+  }
+  const data = await response.json();
+
+  return getApplicableOffers(data);
+}
+
+/**
+ * Get the main element after it is decorated.
+ */
+function getDecoratedContent() {
+  return new Promise((resolve) => {
+    if (document.body.classList.contains('appear')) {
+      // eslint-disable-next-line no-console
+      console.debug('content is already decorated... resolving immediately');
+      resolve(document.body.querySelector('main'));
+    }
+    const config = {
+      attributes: true,
+      attributeFilter: ['class'],
+    };
+    const observer = new MutationObserver(() => {
+      if (document.body.classList.contains('appear')) {
+        // eslint-disable-next-line no-console
+        console.debug('content has been decorated... resolving');
+        observer.disconnect();
+        resolve(document.body.querySelector('main'));
+      }
+    });
+    observer.observe(document.body, config);
+  });
+}
+
+/**
+ * Get all sections that are already loaded.
+ * @param main The main element.
+ */
+function getLoadedSections(main) {
+  const sections = main.querySelectorAll('.section');
+  return Array.from(sections)
+    .map((section) => new Promise((resolve) => {
+      if (section.getAttribute('data-section-status') === 'loaded') {
+        // eslint-disable-next-line no-console
+        console.debug('section is already loaded... resolving immediately', section);
+        resolve(section);
+      }
+      const config = {
+        attributes: true,
+        attributeFilter: ['data-section-status'],
+      };
+      const observer = new MutationObserver(() => {
+        if (section.getAttribute('data-section-status') === 'loaded') {
+          // eslint-disable-next-line no-console
+          console.debug('section has been loaded... resolving', section);
+          observer.disconnect();
+          resolve(section);
+        }
+      });
+      observer.observe(section, config);
```

Comment:
having one observer for each section has a performance impact. not a big one as we only have a few sections, but why not having a single observer on `<main>` instead? That code would be significantly easier to understand.

Reply:
I was unsure about the performance impact, so I am attaching an observer to smaller DOM subtrees.